sdk/communication/azure-communication-common/src/main/java/com/azure/communication/common/CommunicationTokenCredential.java

```diff
@@ -47,22 +46,54 @@ public CommunicationTokenCredential(String token) {
      * If refresh proactively is true, callback function tokenRefresher will be called
      * ahead of the token expiry by the number of minutes specified by
-     * CallbackOffsetMinutes defaulted to ten minutes. To modify this default, call
-     * setCallbackOffsetMinutes after construction
+     * CallbackOffsetMinutes defaulted to ten minutes.
      *
      * @param tokenRefreshOptions implementation to supply fresh token when reqested
      */
     public CommunicationTokenCredential(CommunicationTokenRefreshOptions tokenRefreshOptions) {
-        Supplier<Mono<String>> tokenRefresher = tokenRefreshOptions.getTokenRefresher();
-        Objects.requireNonNull(tokenRefresher, "'tokenRefresher' cannot be null.");
-        refresher = tokenRefresher;
+        Supplier<String> tokenRefresher = tokenRefreshOptions.getTokenRefresherSync();
+        refresher = tokenRefresher != null
+                ? () -> Mono.fromSupplier(tokenRefresher)
+                : tokenRefreshOptions.getTokenRefresher();
+        Objects.requireNonNull(refresher, "'tokenRefresher' cannot be null.");
         if (tokenRefreshOptions.getInitialToken() != null) {
             setToken(tokenRefreshOptions.getInitialToken());
-            if (tokenRefreshOptions.isRefreshProactively()) {
-                OffsetDateTime nextFetchTime = accessToken.getExpiresAt().minusMinutes(DEFAULT_EXPIRING_OFFSET_MINUTES);
-                fetchingTask = new FetchingTask(this, nextFetchTime);
-            }
         }
+        if (tokenRefreshOptions.isRefreshProactively()) {
+            scheduleRefresher();
+        }
+    }
+
+    private void scheduleRefresher() {
+        OffsetDateTime nextFetchTime;
+        if (isTokenExpired(accessToken)) {
+            nextFetchTime = OffsetDateTime.now();
+        } else {
+            OffsetDateTime now = OffsetDateTime.now();
+            long tokenTtlMs = accessToken.getExpiresAt().toInstant().toEpochMilli() - now.toInstant().toEpochMilli();
+            long nextFetchTimeMs = isTokenExpiringSoon()
+                    ? tokenTtlMs / DEFAULT_REFRESH_AFTER_TTL_DIVIDER
+                    : tokenTtlMs - TimeUnit.MILLISECONDS.convert(DEFAULT_EXPIRING_OFFSET_MINUTES, TimeUnit.MINUTES);
+            nextFetchTime = now.plusNanos(TimeUnit.NANOSECONDS.convert(nextFetchTimeMs, TimeUnit.MILLISECONDS));
+        }
+        fetchingTask = new FetchingTask(this, nextFetchTime);
+    }
+
+    private boolean isTokenExpired(AccessToken accessToken) {
+        return accessToken == null || accessToken.isExpired();
+    }
+
+    private boolean isTokenExpiringSoon() {
+        return accessToken == null || OffsetDateTime.now().compareTo(accessToken.getExpiresAt().minusMinutes(DEFAULT_EXPIRING_OFFSET_MINUTES)) > 0;
+    }
+
+    /**
+     * Get Azure core access token from credential
+     *
+     * @return Synchronous call to fetch actual token
+     */
+    public AccessToken getTokenSync() {
```

Comment:
I think we shouldn't have introduced this at all. Other packages depend on `getToken` and there's no reason to change anything about that. The `getToken` methods are not intended to be used by customers so there's no reason to introduce this Sync version.